server/lib/nl/detection/palm_api.py

```diff
@@ -0,0 +1,336 @@
+Output JSON:
+```
+{
+  "PLACES":  ["california"],
+  "METRICS": ["size"],
+  "SIZE": "BIG",
+  "SUB_PLACE_TYPE": "CITY"
+}
+```
+
+Input Sentence: "prevalence of asthma vs. temperature rise in US cities where median age is over 30 and income is under $100K."
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "METRICS": ["prevalence of asthma", "temperature rise"],
+  "SUB_PLACE_TYPE": "CITY",
+  "COMPARE": "COMPARE_METRICS",
+  "COMPARISON_FILTER": [
+    {
+      "COMPARISON_METRIC": "median age",
+      "COMPARISON_OPERATOR": "GREATER_THAN",
+      "VALUE": 30
+    },
+    {
+      "COMPARISON_METRIC": "income",
+      "COMPARISON_OPERATOR": "LESSER_THAN",
+      "VALUE": 100000
+    }
+  ]
+}
+```
+
+Input Sentence: "Blood pressure prevalence in US cities with highest income levels"
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "METRICS": ["blood pressure prevalence"],
+  "SUB_PLACE_TYPE": "CITY",
+  "RANK": "HIGH",
+  "RANKING_FILTER": [
+    {
+      "RANKING_METRIC": "income levels",
+      "RANKING_OPERATOR": "IS_HIGHEST"
+    }
+  ]
+}
+```
+
+Input Sentence: "What are some of the biggest floods in the US?"
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "RANK": "HIGH",
+  "DISASTER_EVENT": "FLOOD"
+}
+```
+
+Do not make up new JSON properties outside of the JSON Schema.
+""")
+
+_SUFFIX = '\n\nIn your response, include just the JSON adhering to the above schema. Do not add JSON keys outside the schema.  Also, explain why you set specific enum values.'
+
+# TODO: Consider tweaking this.
+_TEMPERATURE = 1.0
+
+_CANDIDATE_COUNT = 1
+
+_REQ_DATA = {
+    'prompt': {
+        'context': _PROMPT_CONTEXT_ORIG,
+        'messages': {},
+        'examples': [],
+    },
+    'temperature': _TEMPERATURE,
+    'candidateCount': _CANDIDATE_COUNT,
+}
+
+_SKIP_BEGIN_CHARS = ['`', '*']
+
+
+def call(query: str, history: str, ctr: counters.Counters) -> str:
+  logging.info(f'Doing PaLM API for {query}')
+  req_data = _REQ_DATA.copy()
+
+  if not history:
+    q = 'Convert this sentence to JSON: "' + query + '"'
+  else:
+    q = 'As a follow up to the last sentence, convert this sentence to JSON: "' + query + '"'
+  req_data['prompt']['messages']['content'] = q + _SUFFIX
+
+  for input, output in history:
+    req_data['prompt']['examples'].append({
+        'input': {
+            'content': input
+        },
+        'output': {
+            'content': json.dumps(output)
+        }
+    })
+
+  start_time = time.time()
+  req = json.dumps(req_data)
+  api_key = current_app.config['PALM_API_KEY']
+  r = requests.post(f'{_API_URL_BASE}?key={api_key}',
+                    data=req,
+                    headers=_API_HEADER)
+  resp = r.json()
+  ctr.timeit('palm_api_call', start_time)
+
+  return _parse_response(query, resp, ctr)
+
+
+def _parse_response(query: str, resp: Dict, ctr: counters.Counters) -> Dict:
+  if 'candidates' in resp and resp['candidates']:
+    ans = _extract_ans(resp['candidates'][0]['content'])
```

Comment:
nit: supposedly if ans is empty it most likely means a response format which was not as expected, right? in that case, perhaps return {} and log and error/warning/info?

Reply:
Good call.  Added a few checks.